libraries/rundeck_api_client.rb

```diff
@@ -50,23 +51,61 @@ def send_req(req)
       # TODO: protect against redirect loops
       send_req(Net::HTTP::Get.new(res['location']))
     when Net::HTTPClientError, Net::HTTPServerError
-      # Chef::Log.fatal "CODE: #{res.code} PATH: #{path_from_req(req)} BODY: #{res.body[0..250]}"
       raise res.error!
     end
   end
 
-  def get(path)
-    parse_res(send_req(Net::HTTP::Get.new(path)))
+  def request(klass, path, query: {}, payload: {})
+    path = api_path(path)
+
+    unless query.empty?
+      path = [path, URI.encode_www_form(query)].join('?')
+    end
+
+    req = klass.new(path)
+
+    unless payload.empty?
+      req.body = payload.to_json
+    end
+
+    parse_res(send_req(req))
+  end
+
+  def get(path, query={})
+    request(Net::HTTP::Get, path, query: query)
+  end
+
+  def post(path, payload)
+    request(Net::HTTP::Post, path, payload: payload)
   end
 
-  # def post(path, payload: {})
-  #   req = Net::HTTP::Post.new(path)
-  #   req.set_form_data(payload)
-  #   parse_res(send_req(req))
-  # end
+  def put(path, payload)
+    request(Net::HTTP::Put, path, payload: payload)
+  end
+
+  def post_or_put(path, payload)
+    begin
+      post(path, payload)
+    rescue Net::HTTPServerException => e
+      if e.message == '409 "Conflict"'
+        put(path, payload)
+      else
+        raise e
+      end
+    end
+  end
+
+  def delete(path)
+    request(Net::HTTP::Delete, path)
+  end
 
   def parse_res(res)
-    JSON.parse(res.body)
+    # TODO: make this decision based on Content-Type of response
+    if res.body.to_s.empty? || res.body.to_s.start_with?('<!DOCTYPE html>')
```

Comment:
Why cant we just read the content.  We have the reaponse vs the string check its doing now.  Seems brittle.